From: Liam Howlett <liam.howlett@oracle.com>
Subject: maple_tree: do not inline write slow path
Date: Thu, 21 Jul 2022 00:53:02 +0000

Having the slow path inlined causes too much stack usage.  Create new
function mas_wr_bnode() to write a big node into the tree

Link: https://lkml.kernel.org/r/20220721005237.377987-1-Liam.Howlett@oracle.com
Signed-off-by: Liam R. Howlett <Liam.Howlett@oracle.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 lib/maple_tree.c |   22 +++++++++++++++++-----
 1 file changed, 17 insertions(+), 5 deletions(-)

--- a/lib/maple_tree.c~maple_tree-do-not-inline-write-slow-path
+++ a/lib/maple_tree.c
@@ -4309,12 +4309,27 @@ static inline bool mas_wr_append(struct
 	return false;
 }
 
+/*
+ * mas_wr_bnode() - Slow path for a modification.
+ * @wr_mas: The write maple state
+ *
+ * This is where split, rebalance end up.
+ */
+static void mas_wr_bnode(struct ma_wr_state *wr_mas)
+{
+	struct maple_big_node b_node;
+
+	trace_ma_write(__func__, wr_mas->mas, 0, wr_mas->entry);
+	memset(&b_node, 0, sizeof(struct maple_big_node));
+	mas_store_b_node(wr_mas, &b_node, wr_mas->offset_end);
+	mas_commit_b_node(wr_mas, &b_node, wr_mas->node_end);
+}
+
 static inline void mas_wr_modify(struct ma_wr_state *wr_mas)
 {
 	unsigned char node_slots;
 	unsigned char node_size;
 	struct ma_state *mas = wr_mas->mas;
-	struct maple_big_node b_node;
 
 	/* Direct replacement */
 	if (wr_mas->r_min == mas->index && wr_mas->r_max == mas->last) {
@@ -4350,10 +4365,7 @@ static inline void mas_wr_modify(struct
 		return;
 
 slow_path:
-	memset(&b_node, 0, sizeof(struct maple_big_node));
-	mas_store_b_node(wr_mas, &b_node, wr_mas->offset_end);
-	trace_ma_write(__func__, mas, 0, wr_mas->entry);
-	mas_commit_b_node(wr_mas, &b_node, wr_mas->node_end);
+	mas_wr_bnode(wr_mas);
 }
 
 /*
_
